shared/src/store/action-reducers/utils/spatial-elements.ts

```diff
@@ -0,0 +1,86 @@
+import type { Material, Patient, Personnel } from '../../../models';
+import type { Position } from '../../../models/utils';
+import { SpatialTree } from '../../../models/utils/spatial-tree';
+import type { ExerciseState } from '../../../state';
+import type { Mutable, UUID } from '../../../utils';
+import { cloneDeepMutable } from '../../../utils';
+import { updateTreatments } from './calculate-treatments';
+import { getElement } from './get-element';
+
+/**
+ * The element types for which a spatial tree exists in the state to improve the performance.
+ * The position of the element must be changed via one of the function in this file.
+ * In addition, the respective functions must be called when an element gets added or removed.
+ */
+export type SpatialElementType = 'materials' | 'patients' | 'personnel';
+
+/**
+ * Adds an element with a position and executes side effects to guarantee the consistency of the state.
+ * Must be called if an element is added to the state
+ */
+export function addElementPosition(
```

Comment:
this function is not needed (except for the name or so), as updateElementPosition could do exactly the same or not?

Reply:
The general idea for the API here was: 
* `addElementPosition` must be called when an element is added to the state.
* `updateElementPosition` must be called when the position of an (previously) added element changes
* `removeElementPosition` must be called when an element that was added gets deleted

The knowledge that elements without a position are not in the SpatialTree should be kept internally and not exposed outside this API.

I'll look at these three comments about it again later.

Reply:
I see, but should `addElementPosition` be called when the `Position === undefined` (see here https://github.com/hpi-sam/digital-fuesim-manv/pull/523#discussion_r978822501) ?

Reply:
>

> I see, but should `addElementPosition` be called when the `Position === undefined` (see here [#523 (comment)](https://github.com/hpi-sam/digital-fuesim-manv/pull/523#discussion_r978822501)) ?

According to my API definition: Yes.
> `addElementPosition` must be called when an element is added to the state.

Reply:
Another proposal for an API:

```ts
/**
 * @param previousPosition if an element gets added, this must be `undefined`
 * @param newPosition if an element gets removed, this must be `undefined`
 * Must be called every time the position of an element changes (including if it gets added or removed)
 * Updates the Spatial tree and the element's position property to {@link newPosition}.
 */
function updateElementPosition(state, elementType, elementId, previousPosition: Position | undefined, newPosition: Position | undefined)
```

Reply:
Or: 
```ts
/**
 * @param element the element whose position should be updated (the position property must still be the position before the update - exception: the element just got added and the previous position is the same as the new position)
 * @param elementType TODO: incorporate this into element
 * @param newPosition if an element gets removed, this must be `undefined`
 * Must be called every time the position of an element changes (including if the element gets added or removed)
 * Updates the Spatial tree and the element's position property to {@link newPosition}.
 */
function updateElementPosition(state, element: Element, elementType, newPosition: Position | undefined)
```

Reply:
Something like that the two proposal could be better than to have 3 different functions, you only need to remember "always call this function to change the position of an element".
But having three different functions could also make more clear what is happining. 

What about making it impossible to change the position in any other way than through this (or three) function(s)?
(Using position setter and getter)?

Reply:
Getter and setter do not work in this context. I think a comment `Reducers should only change the position via {@link ...}` is best one could do...

Reply:
@ClFeSc what do you think and more intuitive/logic to use of the two proposals or the one that is right now in the PR (with three functions)?

Reply:
I'd stick with the current three functions (`add`, `update`, `delete`) as it's the most flexible for the future (while not being overly complicated - you could call `update` from `add` for example).

Reply:
@Dassderdie from the three functions and your two proposals which do you prefer?

Three functions could be more flexible, but having only one and knowing you have to call exactly this function and have not to think which one you need right now, could be good as well.
Maybe doing the three functions but giving `update` the option to catch intended deletions or adding and doing them or disallowing this and throwing a reducerError as people should use the correct function for the correct action (update meaning having old and new position being defined).

Reply:
> Maybe doing the three functions but giving update the option to catch intended deletions or adding and doing them or disallowing this and throwing a reducerError as people should use the correct function for the correct action (update meaning having old and new position being defined).

Throwing a `ReducerError` when a function is called with the wrong arguments is defensive programming -> bad practice.

I would leave it as it currently is.

Reply:
Then we will leave it, as it is.
But I would make the inputs consistent (where it is possible/sensible).
Either always directly give the element in or only the elementId (and correctly add or delete potential getElements where these functions are called) and call it always `state` or `draftState`

Reply:
@Dassderdie commited changes, if you agree with them, this is resolved.